File: codes/design/LRU_cache.cpp
#include "../base.h"

class LRUCache {
public:
    LRUCache(int capacity) : capacity_(capacity) {
        
    }
    
    int get(int key) {
        if (m_.find(key) == m_.end()) {
             return -1;
        }

        auto pos = m_[key];
        int val = (*pos)[1];
        l_.erase(pos);
        l_.push_front({key, val});
        m_[key] = l_.begin();

        return val;
    }
    
    void put(int key, int value) {
        if (m_.find(key) == m_.end()) {
            if (l_.size() < capacity_) {
                l_.push_front({key, value});
                m_[key] = l_.begin();                 
            } else {
                auto del_pos = l_.back();
                int del_key = del_pos[0];
                m_.erase(del_key);
                l_.pop_back(); 

                l_.push_front({key, value});
                m_[key] = l_.begin(); 
            }

        } else {
            auto pos = m_[key];
            l_.erase(pos);
            l_.push_front({key, value});
            m_[key] = l_.begin();            
        }
        
    }

private:
    unordered_map<int, list<vector<int>>::iterator> m_;
    list<vector<int>> l_;
    int capacity_;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: codes/design/LRU_cache.py
class DLinkedNode:
    def __init__(self, key = 0, val = 0, prev = None, next = None):
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.m = dict()
        self.head = DLinkedNode(-1, -1)
        self.tail = DLinkedNode(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        

    def get(self, key: int) -> int:
        # print("Get Operating: ", key)
        if key not in self.m:
            return -1

        node = self.m[key]
        self.remove(node)
        self.push_front(node)
        # print(self.m)
        # self.printList()

        return node.val
        

    def put(self, key: int, value: int) -> None:
        # print("Put Operating: ", key)
        if key in self.m:
            node = self.m[key]
            node.val = value
            self.remove(node)
            self.push_front(node)
        else:
            node = DLinkedNode(key, value)
            if len(self.m) < self.capacity:
                self.push_front(node)
            else:
                del_node = self.tail.prev
                # print("Deleting", del_node.val)
                del_key = del_node.key
                self.remove(del_node)
                self.push_front(node)
                del self.m[del_key]


            self.m[key] = node
        # print(self.m)
        # self.printList()
    
    def push_front(self, node):
        node.prev = self.head
        node.next = self.head.next

        self.head.next.prev = node
        self.head.next = node


    def remove(self, node):
        prev = node.prev
        new = node.next

        prev.next = new
        new.prev = prev

    def printList(self):
        ls = []
        cur = self.head
        while cur != self.tail:
            ls.append(cur.val)
            cur = cur.next
        ls.append(cur.val)
        print(ls)

        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)